[authpolicy-v2] docs (#275)
Kuadrant/kuadrant-operator: doc/auth.md

# Kuadrant Auth

A Kuadrant AuthPolicy custom resource:

> 1. Allows it to target Gateway API networking resources such as [HTTPRoutes](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRoute) and [Gateways](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.Gateway), using these resources to obtain additional context, i.e., which traffic workload (HTTP attributes, hostnames, user attributes, etc) to enforce auth.
> 2. Allows to specify which specific subsets of the targeted network resource to apply the auth rules to.

Reviewer: The start of these points reads funny to me. I wonder if the suggested is better. 
```suggestion
1. Targets Gateway API networking resources such as [HTTPRoutes](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRoute) and [Gateways](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.Gateway), using these resources to obtain additional context, i.e., which traffic workload (HTTP attributes, hostnames, user attributes, etc) to enforce auth.
2. Specify which specific subsets of the targeted network resource to apply the auth rules to.
```

Author: I can cope with the suggestion for bullet point 1 (to be ported to https://github.com/Kuadrant/kuadrant-operator/blob/main/doc/rate-limiting.md as well), but 2 needs to be clear that specific subsets is optional.

Reviewer: If point 2 needs to be clear that specific subsets are optional I think both version are not clear. Possible might need to use the word optional in there.

3. Abstracts the details of the underlying external authorization protocol and configuration resources, that have a much broader remit and surface area.

> 4. Supports cluster operators to set overrides (soon) and defaults that govern what can be done at the lower levels.

Reviewer 2: perhaps this should be reflected somewhere else? Wondering if we could have an issue label for AuthPolicy for example and link to the GH issues instead?

Author: We have the same line in the RL docs: https://github.com/Kuadrant/kuadrant-operator/blob/ee2a4b36132ed597aeb47ff5eef354267a867451/doc/rate-limiting.md?plain=1#L8. This is basically our promise to support inherited policy attachment eventually.

Reviewer 2: So I think it would still be good to call out that this supports cluster operators setting a default policy based on the Gateway hierarchy? Yes its not defaults overrides but is is a form of default WDYT?

Author: Changed a bit. Please let me know what you think.

Reviewer 2: +1

1. On incoming request, the gateway checks the matching rules for enforcing the auth rules, as stated in the AuthPolicy custom resources and targeted Gateway API networking objects
2. If the request matches, the gateway sends one [CheckRequest](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/external_auth.proto#envoy-v3-api-msg-service-auth-v3-checkrequest) to the external auth service ("Authorino").
3. The external auth service responds with a [CheckResponse](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/external_auth.proto#service-auth-v3-checkresponse) back to the gateway with either an `OK` or `DENIED` response code.

An AuthPolicy and its targeted Gateway API networking resource contain all the statements to configure both the ingress gateway and the external auth service.

### The AuthPolicy custom resource

#### Overview

The `AuthPolicy` spec includes the following parts:

* A reference to an existing Gateway API resource (`spec.targetRef`)
* Authentication/authorization scheme (`spec.rules`)
* Top-level route selectors (`spec.routeSelectors`)
* Top-level additional conditions (`spec.when`)

> * List of stored named patterns (`spec.patterns`)

Reviewer: Why stored named patterns? Is there a way to list non stored patterns? 
```suggestion
* List of named patterns (`spec.patterns`)
```

The auth scheme specify rules for:
* Authentication (`spec.rules.authentication`)
* External auth metadata fetching (`spec.rules.metadata`)
* Authorization (`spec.rules.authorization`)
* Custom response items (`spec.rules.response`)
* Callbacks (`spec.rules.callbacks`)

> where each auth rule can declare specific `routeSelectors` and `when` conditions for the rule to apply.

Reviewer: Very surprised I spotted this.
```suggestion
Where each auth rule can declare specific `routeSelectors` and `when` conditions for the rule to apply.
```

Author: The lower case was intentional.

I know this is not scientific writing, but it's a common pattern in texts where you have formalisation of concepts with blocks in between sentences.

The sentence that starts with "where" should be read in continuity to "The auth scheme specify rules for..."; otherwise it loses context.

But perhaps the following can please us both 🙂:

```suggestion
Each auth rule can declare specific `routeSelectors` and `when` conditions for the rule to apply.
```

        opa: {…}

    # (optional) customizations to the authorization response
    response:
      # custom denial status and other HTTP attributes for unauthenticated requests.
      unauthenticated: {…}

      # custom denial status and other HTTP attributes for unauhtorized requests.
      unauthorized: {…}

      # custom response items when access is granted.
      success:
        # custom response items wrapped as HTTP headers to be injected in the request
        headers:
          "my-custom-header":
            # one-of: plain, json, wristband
            plain: {…}

        # custom response items wrapped as envoy dynamic metadata
        dynamicMetadata:
          # one-of: plain, json, wristband
          "my-custom-dyn-metadata":
            json: {…}

    # (optional) rules for post-authorization callback requests to external services.
    # triggered regardless of the result of the authorization request.
    callbacks:
      "my-webhook":
        http: {…}
```

Check out the [API reference](reference/authpolicy.md) for a full specification of the AuthPolicy CRD.

## Using the AuthPolicy

> ### Targeting a HTTPRoute networking resource

Reviewer 2: Interesting I think I would have inverted these to follow the hierarchy. IE cover gateway first and HTTPRoute second?

Author: I see targeting the HTTPRoute as the simplest case and therefore the one to explain first, but maybe that's because of how later we explain, when targeting a gateway, what happens in case there are also other policies targeting HTTPRoutes under that gateway. If we keep that example as-is, then the reader needs to know what targeting a HTTPRoute means beforehand.

Should we go then?

**OPTION 1 (current)**
- _Targeting a HTTPRoute_ – focus on the direct effect of the policy over that route, no mention about Gateway policies
- _Targeting a Gateway_ – explain how the policy affects all routes under the gateway EXCEPT the ones that have a more specific policy of its own

**OPTION 2**
- _Targeting a Gateway_ – explain how it affects all HTTPRoutes under the gateway, no mention about HTTPRoute policies
- _Targeting a HTTPRoute_ – explain how the policy directly affects that route AND replaces any other policy targeting gateways the route is under

Reviewer 2: So I don't see it as a big deal and not a blocker to merge. But my preference is prob Option 2 as it (perhaps I am biased) fits how I think about gateway API resources
